core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -566,9 +587,63 @@ public void cancelled(Context context) {
 
           context.addListener(new ServerStreamCancellationListener(), directExecutor());
         }
+
+        private <ReqT, RespT> ServerCallParameters<ReqT, RespT> maySwitchExecutor(
+            final ServerMethodDefinition<ReqT, RespT> methodDef,
+            final ServerStream stream,
+            final Metadata headers,
+            final Context.CancellableContext context,
+            final Tag tag) {
+          final ServerCallImpl<ReqT, RespT> call = new ServerCallImpl<>(
+                  stream,
+                  methodDef.getMethodDescriptor(),
+                  headers,
+                  context,
+                  decompressorRegistry,
+                  compressorRegistry,
+                  serverCallTracer,
+                  tag);
+          final Executor switchingExecutor = executorSupplier.getExecutor(call, headers);
+          if (switchingExecutor != null && defaultExecutor instanceof SerializingExecutor) {
+            ((SerializingExecutor)defaultExecutor).setExecutor(switchingExecutor);
+          }
+          return new ServerCallParameters<>(call, methodDef);
+        }
       }
 
-      wrappedExecutor.execute(new StreamCreated());
+      final class CallHandled extends ContextRunnable {
+        CallHandled() {
+          super(context);
+        }
+
+        @Override
+        public void runInContext() {
+          ServerStreamListener listener = NOOP_LISTENER;
+          ServerCallParameters<?,?> callParameters;
+          if (future.isCancelled()) {
```

Comment:
This should be done within the try-finally, as even if cancelled we need to set the listener (to NOOP).

Reply:
oh, because it was handled in the `lookupMethod()`, but yea I think it is better to put it in `callHandled()` try-finally